#### Name: 

# PHYS 230 Lab Assignment 7

### Wednesday, February 11, 2026: Chapter 4.1-4.3
- Speed and Accuracy

#### Start by importing any packages you need below (feel free to update as you go):

In [107]:
import math
import time
import numpy as np
import matplotlib.pyplot as plt

## Short python snippets (10 pts - 5 pts each)
Below each markdow cell, write a short python snippet to perform the indicated action; you will likely have to use google to find out how to perform some of these actions--this is deliberate on my part. Google and *stack overflow* are your friends.<br>
Add a markdown cell to explain your thinking if this is needed.

**1.** There are 3 ways to change the decimal accuracy that you use:
    - Using “%”:- “%” operator is used to format as well as set precision in python. 
    - Using format():- This is yet another way to format the string for setting precision.
    - Using round(x,n):- This function takes 2 arguments, number and a number of decimal points wanted. 
    
   Look up how to use these three different formatting options, and use all three to print out the number $\pi$ to 3 decimal places. 

In [10]:
print('%.3f' % math.pi)
print('{0:.3f}'.format(math.pi))
print(round(math.pi, 3))

3.142
3.142
3.142


**2.** I have given the book example 4.2 below for The Madelung Constant is a dimnensionless quantity in condensed matter physics giving the total electric potential felt by an atom in a solid.  
- Edit the code so you can measure the time it takes to do the whole calculation. Then, time how long it takes the code to run 100 atoms, 1000 atoms, and 1 million. ($L$ is the number of atoms) 
    - *if you are using VScode:* yes I realize that the cells can time the runtime for you, please use code to time it as well
- Using this information, estimate how long it would take for the program to run for 1 billion terms & justify your value using markdown [do not run the code for 1 billion atoms]. 
- Also comment on the difference in the result as the number of terms increases. The expected value is around -1.74699. 

In [36]:
start = time.time()
L = 100 # number of atoms in the solid 
M = 0.0 # Medelung constant starting value

# loop over all directions in the solid 
for i in range(-L,L+1):
    for j in range(-L,L+1):
        for k in range(-L,L+1):
            if i==0 and j==0 and k==0: #condition for if at origin 
                continue
            elif (i+j+k)%2==0: # condition for if i*j*k position is positive 
                M += 1/np.sqrt(i*i+j*j+k*k)
            else: # condition for if i*j*k position is negative 
                M -= 1/np.sqrt(i*i+j*j+k*k)
end = time.time()
print("M =",M)
print('Execution Time:', end - start)

M = -1.7418198158396654
Execution Time: 19.124046325683594


Execution time by value of L:
T(L = 10) = 0.026691 s
T(L = 100) = 19.124 s
T(L = 1000) and T(L = 1e+6) take too long to run to determine by running the program, and so will be estimated with curve fitting. The terms for intermediary values of L were tested for the purpose of computing a curve of best fit.

To estimate execution times for values of L that were too large to run in a reasonable time frame, a curve was fit to a plot of program execution times for L = 10 to L = 100, with an interval of 5 between each value of L.
Using this curve, we can estimate execution times that cannot reasonably be run:
T(L = 1000) = 25.519 hours
T(L = 1e+6) = 2.8027 billion years
T(L = 1e+9) = 2.8026 octillion years

## Write a script 1 (25 pts)
#### Calculating a factorial  (adapted from textbook)

Write a code to do the following: 
- Write a function `factorial_int` to calculate and print the factorial of a number entered by the user and return as an interger value (see section 2.6)
- Write a second function `factorial_float` to calculate the factorial of the entered number as a floating variable instead. 
- Use each to calculate the factorial of 200 entered by a user. 
- What do you find? Include an explanation of the difference between the two in your markdown cell. 

When your program is complete, add a markdown cell below it to explain your process and program and include any values/comparisons. 

In [69]:
def factorial_int(x):
    factorial = 1
    while x > 0:
        factorial = factorial * x
        x = x - 1
    return(factorial)
def factorial_float(x):
    factorial = 1.
    while x > 0.:
        factorial = factorial * x
        x = x - 1.
    return(factorial)

print(factorial_int(200))
print(factorial_float(200))

788657867364790503552363213932185062295135977687173263294742533244359449963403342920304284011984623904177212138919638830257642790242637105061926624952829931113462857270763317237396988943922445621451664240254033291864131227428294853277524242407573903240321257405579568660226031904170324062351700858796178922222789623703897374720000000000000000000000000000000000000000000000000
inf


factorial_int(200) = 788657867364790503552363213932185062295135977687173263294742533244359449963403342920304284011984623904177212138919638830257642790242637105061926624952829931113462857270763317237396988943922445621451664240254033291864131227428294853277524242407573903240321257405579568660226031904170324062351700858796178922222789623703897374720000000000000000000000000000000000000000000000000
factorial_float(200) = inf

The integer form of the function can handle much larger numbers, since the floating point version will overflow when the input is large enough. The floating point version will also round to a certain number of digits and display it in scientific notation, which can harm precision. However, the floating point version displays a much more compact form of the answer when it doesn't overflow.

## Write a script 2 (22 pts)
####  Calculating a derivative (adapted from textbook)

Using the definition of a derivative below we can use a computer take the derivative of a function $f(x)$ at a point $x$:
$$ \frac{df}{dx} = \lim_{h \to 0} \frac{f(x+h)-f(x)}{h}$$
Actually taking $h$ to zero is not reasonable, but we can at least make $h$ small.


Write a code to do the following:
- Define a function `f1(x)` that returns the value of $x(x-1)$ [this should only take you 2 lines]
- Define a second function `f_deriv(x, h)` that takes in an x and h value and uses your `f1(x)` function to calculate the derivative of the function at the point $x = 1$ using the formula above and $h = 10^{-2}$. Have your function return both the result of the derivative and h (in that order). 
- Put the result in a markdown cell and comment on the accuracy by calculating the actual derivative at $x=1$ yourself (you can do this on your own, I know you can!). 
- Repeat the calculation for $h = 10^{-4}, 10^{-6}, 10^{-8}, 10^{-10}, 10^{-12}, 10^{-14}$. You should see an increase in accuracy and then a decrease. In the markdown cell, include all the results and comment on why this occurs. 

When your program is complete, add a markdown cell below it to explain your process and program and include any values/comparisons. 

In [46]:
def f1(x):
    return(x * (x - 1))
def f_deriv(x, h):
    return(((f1(x + h) - f1(x)) / h), h)

print(f_deriv(1, 1e-14))

(0.9992007221626509, 1e-14)


The known value is f'1(1) = 1.
h = 1e-2: f'1(1) = 1.010000000000001     | Error: 1%
h = 1e-4: f'1(1) = 1.0000999999998899    | Error: 0.010000%
h = 1e-6: f'1(1) = 1.0000009999177333    | Error: 0.000099992%
h = 1e-8: f'1(1) = 1.0000000039225287    | Error: (3.9225e-7)%
h = 1e-10: f'1(1) = 1.000000082840371    | Error: 0.0000082840%
h = 1e-12: f'1(1) = 1.0000889005833413   | Error: 0.0088901%
h = 1e-14: f'1(1) = 0.9992007221626509   | Error: 0.079928%

The reason for the accuracy increasing until h = 1e-8 and then decreasing afterwards is likely due to how floating point values are handled. These values are represented by sums of binary fractions when stored, which necessarily makes them imprecise approximations of values that can't be expressed as sums of binary fractions.

## Write a script 3 (22 pts)
####  Calculating integrals (adapted from textbook)

Like you did with a derivative, we can also use the definition of an integral to evaluate integrals with a computer. If we want to evaluate the integral 
 

$$ I = \int_{-1}^{1} \sqrt{1-x^2} \,dx, $$


we can divide the domain of the integration in to a larger number $N$ of slices of width $h = 2/N$ each. The Riemann definition can then be applied: 
$$ I = \lim_{N \to\infty} \sum_{k=1}^{N} hy_k, $$
where 
$$ y_k = \sqrt{1-x_k^2} \hspace{0.2in} \text{and}\hspace{0.2in} x_k = -1+hk.$$


We just need to make $N$ large to get a good approximation (since we can't in practice take the limit as $N\rightarrow \infty$). The expected value of this integral is $\pi/2$. 

Write a code to do the following: 
- Define a function `f2(x)` that returns the value of $\sqrt{1-x^2}.$  [this should only take you 2 lines]
- Define a second function `f_int(N)` that takes in in an N value and uses your `f2(x)` function to evaluate the integral. Have your function return both the result of the derivative and N (in that order).
- $N = 100$ and compare with the expected result. [perhaps a percent error makes a good comparison!]
- Increase the value of $N$ to make it more accurate. If the program must run in 1 second or less, how accurate can you get? 


When your program is complete, add a markdown cell below it to explain your process and program and include any comparisions. 

In [140]:
def f2(x):
    return(math.sqrt(1 - (x ** 2)))
def f_int(N):
    h = 2 / N
    slices = np.linspace(-1, 1, N).tolist()
    area = 0
    for k in slices:
        area = area + (h * f2(-1 + (h * slices.index(k))))
    return(area, N)

start = time.time()
print(f_int(8500))
end = time.time()
print(end - start)

(1.5707942046053063, 8500)
0.7298064231872559


Known Value: pi / 2
f_int(100) = 1.5691342555492505  |  Error: 0.10581%

The runtime is highly variable, so I picked 8500 as the example N, since that practically never exceeded 1 second of time.
f_int(8500) = 1.5707942046053063 | Error: 0.00013510%